crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -6391,6 +6429,48 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn subscript_tuple() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            "
+            t = (1, 'a')
+
+            a = t[0]
+            b = t[1]
+            c = t[2]
+            ",
+        )?;
+
+        assert_public_ty(&db, "/src/a.py", "a", "Literal[1]");
+        assert_public_ty(&db, "/src/a.py", "b", "Literal[\"a\"]");
+        assert_public_ty(&db, "/src/a.py", "c", "Unknown");
+        Ok(())
+    }
+
+    #[test]
+    fn subscript_string() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            "
+            s = 'abc'
+
+            a = s[0]
+            b = s[1]
+            c = s[()]
+            ",
+        )?;
+
+        assert_public_ty(&db, "/src/a.py", "a", "Literal[\"a\"]");
+        assert_public_ty(&db, "/src/a.py", "b", "Literal[\"b\"]");
+        assert_public_ty(&db, "/src/a.py", "c", "LiteralString");
```

Comment:
`c` leads to a `TypeError` at runtime; this isn't correct. We should emit a diagnostic and infer `Unknown` rather than `LiteralString`

Reply:
What is the plan for emitting those diagnostics? Don't we need to propagate some information about _why_ it failed?

Reply:
Of course! Like I said in https://github.com/astral-sh/ruff/pull/13559#issuecomment-2382679447, take a look at the `add_diagnostic` and `not_iterable_diagnostic` methods for how we do this elsewhere